File: MPI/03.Ring/ring.cpp
#include <mpi.h>
#include <vector>

void copyTo(std::vector<int> &dst, std::vector<int> &src){
  for(int i = 0; i < src.size(); i++){
    dst.at(i) = src[i];
  }
}

int main(int argc, char* argv[]){
  int my_rank, num_procs, sum = 0;

  MPI_Init(&argc, &argv);
  MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
  MPI_Comm_size(MPI_COMM_WORLD, &num_procs);

  std::vector<int> rcv_buf = { 0 };
  std::vector<int> snd_buf = { my_rank };
  int right = (my_rank+1) % num_procs;
  int left = (my_rank-1+num_procs) % num_procs;
  MPI_Request request;
  MPI_Status status;

  for(int i = 0; i < num_procs; i++){
    MPI_Irecv(&rcv_buf[0], 1, MPI_INT, left, 17, MPI_COMM_WORLD, &request);
    MPI_Send(&snd_buf[0], 1, MPI_INT, right, 17, MPI_COMM_WORLD);
    MPI_Wait(&request, &status);
    copyTo(snd_buf, rcv_buf);
    sum += rcv_buf[0];
  }

  printf("PE%d:\tSum = %d\n", my_rank, sum);

  MPI_Finalize();
}

File: MPI/03.Ring/ring.py
from mpi4py import MPI
import numpy as np

rcv_buf = np.empty((), dtype=np.intc)
status = MPI.Status()

comm_world = MPI.COMM_WORLD
my_rank = comm_world.Get_rank()
size = comm_world.Get_size()

right = (my_rank+1)      % size;
left  = (my_rank-1+size) % size;

sum = 0
snd_buf = np.array(my_rank, dtype=np.intc)

for i in range(size):
  request = comm_world.Isend((snd_buf, 1, MPI.INT), dest=right, tag=17)
  comm_world.Recv((rcv_buf, 1, MPI.INT), source=left, tag=17, status=status)
  request.Wait()
  np.copyto(snd_buf, rcv_buf)
  sum += rcv_buf
print(f"PE{my_rank}:\tSum = {sum}")